### DISCUSSION TITLE
External control using python

### DISCUSSION THREAD

**Discussion by spicyacorn** [ggarra13/mrv2#322] [Q&A] [answered]
Hello!

I'm trying to control mrv2 from another process, 
which is a python application that allows users to select specific shots/versions to review.

I used KeyframePro in the past which uses exposes an IP socket to allow the player to load new media, read or set the current frame, set the playing state and a few other convenience functions.

Is there a similar API available for mrv2 ?
I can launch a new instance of mrv with a specific OTIO as source using the command line, but am unable to communicate with a running instance.

**Comment by ggarra13** [comment, marked as answer]
> Is there a similar API available for mrv2 ? I can launch a new instance of mrv with a specific OTIO as source using the command line, but am unable to communicate with a running instance.

Congratulations!  You are smart guy.  There's not one, but three different ways in which you can achieve this, depending on your level of knowledge.

1) Simplest.  No knowledge required.  Go to Preferences->Use ComfyUI Pipe->Yes and it will open a port (don't remember the number off the top of my head and it currently cannot be changed).  If you are on Windows, you will also have to deal with Windows' Defender to open a hole in the firewall, with the possible danger it implies.  If you are on macOS or Linux and you are communicating on the same machine no hole in the firewall is needed, only if you communicate remotely.   With this, you can send new media in, but not control mrv2.

2) More powerful.  Needs python knowledge.  Run a python plug-in that starts a new thread, calls Fl.check() or cmd.update() and uses time.time() to give it time for the thread and mrv2's event loop to run.  You can then use the python api to control it a as you see fit.

3) Most powerful.  Needs C++ and, possibly, python knowledge.  Go to the Network Panel and open a port where you see fit.  Again, on Windows, you will need to bypass the firewall.  You can communicate with mrv2 by sending commands through that port.  Commands are sent with:
       1) json compressed as bson for performance.
       2) a simple cypher is used to obfuscate the information.
   You may need to look at the source code in the mrvNetwork directory:  mrvCypher.cpp,  mrvClient.cpp  and mrvCommandInterpreter.cpp to see the list of all commands (and parameters) supported.
